hook-consumer/src/consumer.rs

```diff
@@ -199,12 +192,20 @@ impl<'p> WebhookConsumer<'p> {
 
     /// Run this consumer to continuously process any jobs that become available.
     pub async fn run(&self) -> Result<(), WebhookConsumerError> {
+        let semaphore = Arc::new(sync::Semaphore::new(self.max_concurrent_jobs));
+
         loop {
-            // TODO: The number of jobs processed will be capped by the PG connection limit when running in transactional mode.
             let webhook_job = self.wait_for_job().await?;
 
-            let request_timeout = self.request_timeout; // Required to avoid capturing self in closure.
-            tokio::spawn(async move { process_webhook_job(webhook_job, request_timeout).await });
+            // reqwest::Client internally wraps with Arc, so this allocation is cheap.
+            let client = self.client.clone();
+            let permit = semaphore.clone().acquire_owned().await.unwrap();
+
+            tokio::spawn(async move {
+                let result = process_webhook_job(client, webhook_job).await;
+                drop(permit);
+                result.expect("webhook processing failed");
```

Comment:
Doesn't `expect` on an Error panic? If I'm reading right, this will Error if a PG query fails, but I don't think we want to crash the process in that case do we?

Reply:
Should be fine to return an error too, we are not doing any handling anyways (and that's something that we should, but thought about focusing on tracking/graceful handling afterwards).

Reply:
> Yeah, it will panic, but inside a task (thread), which should not propagate to the main thread: [users.rust-lang.org/t/panic-in-tokio-task-does-not-end-the-program-execution/45731/2](https://users.rust-lang.org/t/panic-in-tokio-task-does-not-end-the-program-execution/45731/2)

Ahhh, TIL, thanks.

Do we benefit somehow from panicking the task? Returning an error just seems easier for future readers (me, lol).

Reply:
Handling gets tricky as if we get to that point it means something is wrong with the database...

Reply:
> Handling gets tricky as if we get to that point it means something is wrong with the database...

That's true, but as currently written we'd drop the txn (or try, it might have disconnected which would do the same thing), release the row, and it would (eventually) retry when the DB worked again right? So everything would stall (as expected) but we wouldn't really lose anything, right? Seems like the behavior we want?